### PR TITLE
fix(v2): enable scrolling for sidebar menu only

### LINKED ISSUE #2620: Sidebar scrolling thumb is not always visible when hideOnScroll = false
## 🐛 Bug Report

The sidebar has some problems with scrolling when hideOnScroll = false. The scrolling thumb is only partially visible when scrolled to top, as it is hidden by the NavBar. If the sidebar content is long enough, the thumb can disappear completely behind the NavBar.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

## To Reproduce

1. Go to our documentation site: support.touchgfx.com
2. Expand enough categories in the sidebar such that the scrollbar appears
3. Part of the scrollbar is now hidden behind the NavBar.

## Expected behavior

I expect to see the whole scrollbar when scrolled all the way to the top in the sidebar and on the page.

## Actual Behavior

Part of the scrollbar is hidden. The whole scrollbar is hidden if the content of the sidebar is long enough. 

hidden-scrollbar-bug

## Your Environment

- Docusaurus version used: v. 2.0.0-alpha.50
- Version 78.0.3904.97 (Official Build) (32-bit)

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

This PR resolves #2620 and includes complex changes, mainly due to support for announcement bar. Since the height of the sidebar is based on the full height of viewport, when we are at the very top of the page, we see an announcement bar, which has a certain height, which should be taken into account when calculating the height of the sidebar.

In the screenshots below are full reflection of this challenge, for the solution of which we need to expose announcement bar info (to find out whether the announcement bar was dismissed or not) and limit the height of the announcement bar to a fixed height on the desktop.

Without announcement bar, sidebar menu scrolling is fully visible and everything is fine:

снимок

_But_ if the announcement bar is shown, then bottom part of the sidebar menu scrolling is no longer visible, since we need to **somehow** take into account the height of this bar. I think it's not so easy to do.

снимок_1

To solve that, I created Provider+Context for the announcement bar, so that I could know from the DocSidebar component whether the bar was dismissed, and if not, then "adjust" the height of  sidebar menu so that scrolling sidebar menu looks as expected.
As a result, we have the desired result! However how much code was needed to solve this!🤕

снимок_2

* * *

This PR contains two commits to show how much effort it took to get things to work with the _enabled_ announcement bar.
Enough changes from the first commit https://github.com/facebook/docusaurus/commit/20bc7cc9911beadad9d564a263b1dd684a8373f1 however we will have new issue with a scrolling of sidebar menu (as shown in the second screenshot).
We definitely do not want this, but I could not find an _easier_ solution, maybe there will be thoughts on how to simplify this (@yangshun maybe I missed something?)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Preview

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

**Comment by lex111** [comment]
FYI in Next.js docs has such a problem with sidebar menu height right now  https://nextjs.org/docs/getting-started

screencast-nextjs org-2020 05 07-11_51_18


I marked the empty bar with a red arrow - this issue solved in my solution, so there is so much code to fix that (we need to recalculate menu height after scroll of announcement bar).

**Comment by lex111** [comment]
@yangshun any progress on this PR?

**Comment by lex111** [comment]
In fact, the main changes are regarding the addition of new context/provider for announcement bar, so we can combine this context with the context for the syncing tab choices feature. For example, it could be something like `UserSettingsProvider`.

Yes, this is not a very serious bug, but this does not mean that we do not need to fix it.
You do not take into account the use cases when the docs categories are expanded by default. Beside this fact, the more people look at the documentation on tablets, so this issue is visible even if there are not many docs.

We cannot refuse to fix this issue just because the announcement bar is not such a popular feature. This fix applies not only to the announcement bar (I added support for this feature to avoid the problem as on Next.js docs).
I understand the desire to reduce the amount of  needed code, so as I suggested above, we can combine the provider (`TabGroupChoiceProvider` and `AnnouncementBarProvider` -> `UserSettingsProvider`) into one.

**Comment by lex111** [comment]
Fixed now.

**Comment by lex111** [comment]
cc @yangshun

File: packages/docusaurus-theme-classic/src/theme/hooks/useAnnouncementBar.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+const STORAGE_ID_KEY = 'docusaurus.announcement.id';
+
+const useAnnouncementBar = () => {
+  const {
+    siteConfig: {
+      themeConfig: {
+        announcementBar: {id},
+      },
+    } = {},
+  } = useDocusaurusContext();
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/useAnnouncementBar.js:21]: Subjective, but i'd rather use something like `useDocusaurusContext().siteConfig.themeConfig?. announcementBar?.id`

**Pull request merged**